### ISSUE TITLE
hash

### ISSUE THREAD

**Issue by schackbrian2012** [racket/typed-racket#605]
### What version of Racket are you using?
6.10.1

### What program did you run?
```
#lang typed/racket
(hash 0 1 2 3 4 5 6 7 8 9)
```

### What should have happened?
`'#hash((2 . 3) (6 . 7) (0 . 1) (4 . 5) (8 . 9))`

### If you got an error message, please include it here.
```
Type Checker: Polymorphic function `hash' could not be applied to arguments:
Types: a b a b a b a b #f * -> (Immutable-HashTable a b)
       a b a b a b #f * -> (Immutable-HashTable a b)
       a b a b #f * -> (Immutable-HashTable a b)
       a b #f * -> (Immutable-HashTable a b)
       #f * -> (Immutable-HashTable a b)
Arguments: Zero One Positive-Byte Positive-Byte Positive-Byte Positive-Byte Positive-Byte Positive-Byte Positive-Byte Positive-Byte #f *
Expected result: AnyValues
 in: (hash 0 1 2 3 4 5 6 7 8 9)
```

**Comment by pnwamk** [comment]
There is not a type in Typed Racket's vocabulary currently that can describe the type of `hash` (it would require describing rest arguments with a non-uniform type (e.g. the key type followed by the value type followed by the key type etc...).

You can, however, write out the sequence of key/value pairs as a list of cons cells and use `make-hash`.

**Comment by pnwamk** [comment]
or `make-immutable-hash` (depending on whether you wanted a mutable or imutable hash)

**Comment by schackbrian2012** [comment]
@pnwamk Thank you, in this situation, my keys and values were constants, so I used #hash. But this is still a missing and undocumented functionality in typed/racket.

**Comment by pnwamk** [comment]
The type that was printed out documents what `hash` supports (perhaps there are better ways to inform users of the supported type?). In general though, there will be functions where the supported type is not as expressive as the original function, and in those cases you'll have to either decipher what the type means, or ask ;-)

**Comment by pnwamk** [comment]
BTW, [:print-type](https://docs.racket-lang.org/ts-reference/Exploring_Types.html?q=%3Atype#%28form._%28%28lib._typed-racket%2Fbase-env%2Fprims..rkt%29._~3aprint-type%29%29) may be the easiest way to see what Typed Racket thinks the type of an identifier is (if a type exists)

**Comment by samth** [comment]
I think we should just add a special case to do better here.

**Comment by pnwamk** [comment]
for `hash`? syntactically recognize it and special case?

**Comment by samth** [comment]
Yes (and `hasheq` and `hasheqv`).

**Comment by schackbrian2012** [comment]
@pnwamk I noticed that the Typed Racket Guide has a "Caveats and Limitations" section [0]. Would it be possible to list the functions and forms that are missing or not as expressive as in untyped Racket? When I originally did :print-type on the hash function, I saw that it only listed some of the possible types, so I assumed that the list of types must be abbreviated. Then I tried to use it, and it didn't work. I thought that I had made a mistake. When I eventually narrowed it down, I was surprised to find that the bug was not in my code at all.

[0] file:///Applications/Racket%20v6.10.1/doc/ts-guide/caveats.html

@samth Thank you for reopening.
